Simd (#13262)
dotnet/docs: docs/standard/simd.md

---
title: SIMD in .NET
description: This tutorial demonstrates how to use hardware SIMD operations in C# and .NET Core.
author: FIVIL
ms.author: tagoo
ms.date: 03/31/2020
ms.technology: dotnet-standard
---

# Overview

SIMD (Single instruction, multiple data) provides hardware support for performing an operation on multiple pieces of data, in parallel, using a single instruction. In .NET, there's set of SIMD-accelerated types under the <xref:System.Numerics> namespace. SIMD operations can be parallelized at the hardware level. That increases the throughput of the vectorized computations, which are common in mathematical, scientific, and graphics apps.

## .NET SIMD-accelerated types

The .NET SIMD-accelerated types include the following types:

- The <xref:System.Numerics.Vector2>, <xref:System.Numerics.Vector3>, and <xref:System.Numerics.Vector4> types, which represent vectors with 2, 3, and 4 <xref:System.Single> values.

> - Two matrix types, <xref:System.Numerics.Matrix3x2>, which represents a 3x2 matrix, and <xref:System.Numerics.Matrix4x4>, which represents a 4x4 matrix.

Reviewer: Should we call out that this is xref:System.Single values, like we did for Vector2/3/4?

Likewise for Plane/Quaternion?

Reviewer 2: Would you be willing to write that content? It's a bit out of my area of expertise.

Reviewer: ```suggestion
- Two matrix types, <xref:System.Numerics.Matrix3x2>, which represents a 3x2 matrix, and <xref:System.Numerics.Matrix4x4>, which represents a 4x4 matrix of <xref:System.Single> values.
```

- The <xref:System.Numerics.Plane> type, which represents a plane in three-dimensional space.

- The <xref:System.Numerics.Quaternion> type, which represents a vector that is used to encode three-dimensional physical rotations.

> - The <xref:System.Numerics.Vector%601> type, which represents a vector of a specified numeric type and provides a broad set of operators that benefit from SIMD support. The count of a <xref:System.Numerics.Vector%601> instance is fixed, but its value <xref:System.Numerics.Vector%601.Count%2A?displayProperty=nameWithType> depends on the CPU of the machine running the code.

Reviewer: This could probably be reworded to better indicate it is fixed for the lifetime of an application but the count is dependent on the underlying hardware.

Reviewer: ```suggestion
- The <xref:System.Numerics.Vector%601> type, which represents a vector of a specified numeric type and provides a broad set of operators that benefit from SIMD support. The count of a <xref:System.Numerics.Vector%601> instance is fixed for the lifetime of an application, but its value <xref:System.Numerics.Vector%601.Count%2A?displayProperty=nameWithType> depends on the CPU of the machine running the code.
```

  > [!NOTE]
  > The <xref:System.Numerics.Vector%601> type is not included in the .NET Framework. You must install the [System.Numerics.Vectors](https://www.nuget.org/packages/System.Numerics.Vectors) NuGet package to get access to this type.
  
The SIMD-accelerated types are implemented in such a way that they can be used with non-SIMD-accelerated hardware or JIT compilers. To take advantage of SIMD instructions, your 64-bit apps must be run by the runtime that uses the **RyuJIT** compiler. A **RyuJIT** compiler is included in .NET Core and in .NET Framework 4.6 and later. SIMD support is only provided when targeting 64-bit processors.

## How to use SIMD?

> Before executing custom SIMD algorithms, it's possible to check if the host machine supports SIMD by using <xref:System.Numerics.Vector.IsHardwareAccelerated?displayProperty=nameWithType>, which returns a <xref:System.Boolean>. This doesn't guarantee that SIMD-acceleration is enabled for a specific type, but is an indicator that it's supported by some types.

Reviewer: Maybe add in that it doesn't guarantee for a specific type or even for a particular method of a supported type?

That is, `Vector<float>` could be partially accelerated and so `Vector<float>.Method1` might be accelerated while `Vector<float>.Method2` might not be.
It is all dependent on the underlying hardware.

```csharp
var m1 = new Matrix4x4(
            1.1f, 1.2f, 1.3f, 1.4f,
            2.1f, 2.2f, 3.3f, 4.4f,
            3.1f, 3.2f, 3.3f, 3.4f,
            4.1f, 4.2f, 4.3f, 4.4f);

var m2 = Matrix4x4.Transpose(m1);
var mResult = Matrix4x4.Multiply(m1, m2);
```

## Vector\<T>

The <xref:System.Numerics.Vector%601> gives the ability to use longer vectors. The count of a <xref:System.Numerics.Vector%601> instance is fixed, but its value <xref:System.Numerics.Vector%601.Count%2A?displayProperty=nameWithType> depends on the CPU of the machine running the code.

The example below demonstrates adding long arrays elements using <xref:System.Numerics.Vector%601>.

```csharp
double[] SimdVectorProd(double[] left, double[] right)
{
    var offset = Vector<double>.Count;
    double[] result = new double[left.Length];
    int i = 0;
    for (i = 0; i < left.Length; i += offset)
    {
        var v1 = new Vector<double>(left, i);
        var v2 = new Vector<double>(right, i);
        (v1 * v2).CopyTo(result, i);
    }

    //remaining items
    for (; i < left.Length; ++i)
    {
        result[i] = left[i] * right[i];
    }

    return result;
}
```

## Remarks

> SIMD is more likely to remove one bottleneck and expose the next, for example memory throughput. In general the performance benefit of using SIMD varies depending on the specific scenario, and in some cases it can even perform worse than simpler non-SIMD equivalent code.

Reviewer: End with a note that users should benchmark and profile their code to identify hot spots and ensure that changes are meeting their performance goals/needs?

Can we link to https://github.com/dotnet/BenchmarkDotNet?

---
title: SIMD in .NET
description: This tutorial demonstrates how to use SIMD in C#.
author: FIVIL
#ms.author: [MICROSOFT ALIAS OF INTERNAL OWNER]
ms.date: 07/05/2019
#ms.topic: [TOPIC TYPE]
#ms.prod: [PRODUCT VALUE]
ms.technology: dotnet-standard
---

# Overview

> SIMD (Single instruction, multiple data) provides the functionality to do multiple processes on a single core to achieve peak performance. In .NET there is set of SIMD-enabled types under <xref:System.Numerics> namespace. SIMD operations can be parallelized at the hardware level. That increases the throughput of the vectorized computations, which are common in mathematical, scientific, and graphics apps.

Reviewer: This first sentence could be reworded.

Maybe something like:
> SIMD (Single instruction, multiple data) provides hardware support for performing an operation on multiple pieces of data, in parallel, using a single instruction.

> SIMD (Single instruction, multiple data) provides the functionality to do multiple processes on a single core to achieve peak performance. In .NET there is set of SIMD-enabled types under <xref:System.Numerics> namespace. SIMD operations can be parallelized at the hardware level. That increases the throughput of the vectorized computations, which are common in mathematical, scientific, and graphics apps.

Reviewer: It is probably worthwhile to callout both the `System.Numerics` types and the System.Runtime.Intrinsics namespaces. 
The former provides SIMD accelerated functions (in `Vector2/3/4` and other types) and more direct SIMD access (`Vector<T>`) which may have a software fallback on some platforms (such as `arm32`).

The latter provides direct hardware access and no software fallback; so support must be explicitly queried and fallbacks need to be provided by the developer.

Author: Actually, I have just started to work with new `System.Runtime.Intrinsic` in .net core 3 and I will complete these documents based on this [blog post](https://devblogs.microsoft.com/dotnet/hardware-intrinsics-in-net-core/) of yours, very soon

> SIMD (Single instruction, multiple data) provides the functionality to do multiple processes on a single core to achieve peak performance. In .NET there is set of SIMD-enabled types under <xref:System.Numerics> namespace. SIMD operations can be parallelized at the hardware level. That increases the throughput of the vectorized computations, which are common in mathematical, scientific, and graphics apps.

Reviewer: ```suggestion
SIMD (Single instruction, multiple data) provides hardware support for performing an operation on multiple pieces of data, in parallel, using a single instruction. In .NET there is set of SIMD-enabled types under <xref:System.Numerics> namespace. SIMD operations can be parallelized at the hardware level. That increases the throughput of the vectorized computations, which are common in mathematical, scientific, and graphics apps.
```

## .NET SIMD-enabled types

> The .NET SIMD-enabled types include the following:

Reviewer: I'm not sure saying they are `SIMD-enabled` is correct. Its that certain exposed APIs on the types may be SIMD accelerated.

> The .NET SIMD-enabled types include the following:

Reviewer: ```suggestion
## .NET SIMD accelerated types

The .NET SIMD accelerated types include the following:
```

- The <xref:System.Numerics.Vector2>, <xref:System.Numerics.Vector3>, and <xref:System.Numerics.Vector4> types, which represent vectors with 2, 3, and 4 <xref:System.Single> values.

- Two matrix types, <xref:System.Numerics.Matrix3x2>, which represents a 3x2 matrix, and <xref:System.Numerics.Matrix4x4>, which represents a 4x4 matrix.

> - The <xref:System.Numerics.Plane> type, which represents a plane in three-dimensional space.

Reviewer: ```suggestion
- The <xref:System.Numerics.Plane> type, which represents a plane in three-dimensional space using <xref:System.Single> values.
```

> - The <xref:System.Numerics.Quaternion> type, which represents a vector that is used to encode three-dimensional physical rotations.

Reviewer: ```suggestion
- The <xref:System.Numerics.Quaternion> type, which represents a vector that is used to encode three-dimensional physical rotations using <xref:System.Single> values.
```

- The <xref:System.Numerics.Vector%601> type, which represents a vector of a specified numeric type and provides a broad set of operators that benefit from SIMD support. The count of a <xref:System.Numerics.Vector%601> instance is fixed, but its value <xref:System.Numerics.Vector%601.Count%2A?displayProperty=nameWithType> depends on the CPU of the machine, on which code is executed.
  > [!NOTE]
  > The <xref:System.Numerics.Vector%601> type is not included into the .NET Framework. You must install the [System.Numerics.Vectors](https://www.nuget.org/packages/System.Numerics.Vectors) NuGet package to get access to this type.
  
The SIMD-enabled types are implemented in such a way that they can be used with non-SIMD-enabled hardware or JIT compilers. To take advantage of SIMD instructions, your 64-bit apps must be run by the runtime that uses the RyuJIT compiler, which is included in .NET Core and in the .NET Framework 4.6 and later versions. It adds SIMD support when targeting 64-bit processors.

## How to use SIMD?

>  Before executing custom SIMD algorithms, it is possible to check if the host machine supports SIMD using <xref:System.Numerics.Vector.IsHardwareAccelerated?displayProperty=nameWithType>, which returns a <xref:System.Boolean>.

Reviewer: I think this should be explained that `IsHardwareAccelerated` returning `true` only indicates that some functionality may be hardware accelerated. It doesn't provide any guarantees of what functions are accelerated (you may even end up in a scenario where one function is accelerated but another function is not).

Author: This makes sense, I will make some changes.
@tannergooding  Tnx for your suggestions, I will make some changes based on your review

>  Before executing custom SIMD algorithms, it is possible to check if the host machine supports SIMD using <xref:System.Numerics.Vector.IsHardwareAccelerated?displayProperty=nameWithType>, which returns a <xref:System.Boolean>.

Reviewer: ```suggestion
 Before executing custom SIMD algorithms, it is possible to check if the host machine supports SIMD using <xref:System.Numerics.Vector.IsHardwareAccelerated?displayProperty=nameWithType>, which returns a <xref:System.Boolean>. However, this does not indicate all members are accelerated just that some members may be. Which members are accelerated are hardware dependent.
```
